### PR TITLE
Wait for cluster state in recovery

### PR DESCRIPTION
**Pull request by DaveCTurner**
A peer recovery is triggered by the application of a cluster state on
the target node, which creates a local `IndexShard` instance and then
reaches out to the source node to coordinate the recovery. If the source
node has not applied the same cluster state yet then it will respond
with a `DelayRecoveryException` which tells the target node to wait a
little and then retry.

With this commit we now wait on the source node until a fresh enough
cluster state is applied, avoiding the need for the timed wait on the
target node (and another network round trip) before the recovery can
begin.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
WIP (no tests of the new behaviour yet) - just seeing if CI finds any problems

**Review by DaveCTurner — COMMENTED** [review]
Most of the changes in this PR are just plumbing to make the cluster state version available to the `PeerRecoverySourceService`. I've called out the actual change in behaviour, and the test, in comments below.

File: server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceClusterStateDelay.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+public class PeerRecoverySourceClusterStateDelay {
+    private PeerRecoverySourceClusterStateDelay() {}
+
+    private static final Logger logger = LogManager.getLogger(PeerRecoverySourceClusterStateDelay.class);
+
+    /**
+     * Waits for the given cluster state version to be applied locally before proceeding with recovery
+     */
+    public static <T> void ensureClusterStateVersion(
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceClusterStateDelay.java:31]: This is the place that implements the waiting. Extracted to a utility class so we can use it in Serverless too.

File: server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java [hunk 2]
```diff
@@ -1614,6 +1627,127 @@ public void testReservesBytesDuringPeerRecoveryPhaseOne() throws Exception {
         );
     }
 
+    public void testWaitForClusterState() throws Exception {
```
Comment (fcofdez) [inline server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java:1630]: nit: maybe we should rename this test to `testWaitForClusterStateToBeAppliedInSourceNode`?

File: server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java [hunk 3]
```diff
@@ -1614,6 +1627,127 @@ public void testReservesBytesDuringPeerRecoveryPhaseOne() throws Exception {
...
+                final var primaryProceedListener = clusterStateSyncListeners.getStateApplyDelayListener(event.state().version());
+                final var indexRoutingTable = event.state().routingTable().index(indexName);
+                assertNotNull(indexRoutingTable);
+                final var indexShardRoutingTable = indexRoutingTable.shard(0);
+                if (indexShardRoutingTable.size() == 2 && indexShardRoutingTable.getAllInitializingShards().isEmpty() == false) {
+                    // this is the cluster state update which starts the recovery, so delay the primary node application until recovery
+                    // has started
+                    clusterStateSyncListeners.delayUntilRecoveryStart(primaryProceedListener);
+                } else {
+                    primaryProceedListener.onResponse(null);
```
Comment (fcofdez) [inline server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java:1740]: shouldn't this be called once the recovery starts through `clusterStateSyncListeners.onStartRecovery()`?

**Review by fcofdez — APPROVED** [review]
LGTM, I left a question in the test, otherwise this looks good 👍. Sorry for the delay reviewing this.

File: server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java:1739]: This branch is for cluster state updates that _aren't_ the one which starts the update, particularly ones before the recovery, and we need to let the primary proceed with those updates straight away. I added a comment in e9ad47f77225c8a3acafeb7b34a078ffe2d2e589 which I hope helps.

File: server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java]: ++ done in 55ace3e1aafb840cbe2647884f04b1e6055221d3

**Pull request merged**